### ISSUE TITLE
Rocket attacks and Scrambling resolved during wrong phases

### ISSUE THREAD

**Issue by panther2** [triplea-game/triplea#3846]
### Engine version
Tested with 11440, but this is not a new issue.

### Map name

wwII_global1940

### Can you describe how to trigger the error? (eg: what sequence of actions will recreate it?)

TripleA incorrectly resolves rolling for rockets during Combat Move Phase.
TripleA incorrectly resolves scrambling during Conduct Combat Phase,

### Do you have the exact error text? Please copy/paste if so

An image instead:
![rocket](https://user-images.githubusercontent.com/20812775/44306675-08a02380-a394-11e8-8a61-2e1667279af0.PNG)

### Instead of this error, what should have happened?

Scrambling should occur at the very end of the Combat Move Phase.
Rocket Attacks should occur at the beginning of the Conduct Combat Phase (by the rules "during SBR", but it has been clarified that it can occur before SBR).

### Any additional information that may help

In a game without technology the issue does not appear to be relevant.
But in games with technology enabled the issue has illegal side effects:
For example a rocket attack can prevent air units from scrambling.

If resolved correctly, the air units would be able to scramble (at the end of the Combat Move Phase) before the Airfield becomes inoperative due to rocket attacks (at the beginning of the Coduct Combat Phase).
[ScrambleRocket.zip](https://github.com/triplea-game/triplea/files/2300157/ScrambleRocket.zip)

**Comment by Cernelius** [comment]
Oh, I noticed only now that in the image you posted the rockets were made before some combat moves on the same phase. This is surprising to me, as I remember that they used to happen at the end of Combat Move, after clicking on "Done". I'll have to check that out.

**Comment by Cernelius** [comment]
Disregarding the intended/official rules for a moment, I believe that it makes more sense to decide where you are attacking with your rockets at the end of the Combat Move phase, then resolving it at the start of the Combat phase, as this is what you are substantially doing with your bombers (deciding where the bombers are bombing during Combat Move and bombing at start of Combat).

**Comment by Cernelius** [comment]
I've just tested "World War II v3 1941", editing Rockets Advance for Russians, and the rockets were decided and resolved at the end of the Combat Move phase, after having finalized all moves, by clicking on "Done". I believe that reported behaviour of making rockets attack before any combat movements is not present (not possible) in this game (nor in my mods of it).

**Comment by Cernelius** [comment]
From pos2:

`fireRockets 				= true/false (default depends on game option [ww2v2 and ww2v3 have this after delegates that end with "CombatMove" but not "NonCombatMove"])`

So if in that game you can do any combat movements after having fired any rockets, as per your image, that is an unintended behaviour, according to pos2 (assuming it is correct and exhaustive). I don't know why that would happen.

**Comment by panther2** [comment]
@Cernelius 

> What do you mean? Can you link to those clarifications? Can you still decide to make any bombing raids before some rocketing, if you so prefer?

Rulebook Europe/Pacific 1940 2ndEd.:

> Rockets. Your air bases can now launch rockets. **During the Strategic and Tactical Bombing Raids step of your Conduct Combat** phase each turn, each of your operative air bases can make a single rocket attack against an enemy industrial complex, air base, or naval base within 4 spaces of it. This attack does one die roll of damage to that facility. Rockets may not be fired over neutral territories.

Clarification starting from here:
https://www.axisandallies.org/forums/index.php?topic=28562.msg1632681#msg1632681
and especially:
https://www.axisandallies.org/forums/index.php?topic=28562.msg1648393#msg1648393

The written sequence in the above image is a bit irritating. Indeed rockets fire after any movement has been done. That is why  I provided a savegame scenario to test it.

The main problem I see is that scrambling needs to take place before rocket attacks and SBR occurs.

Strictly from the TripleA game mechanics I see no problems when rockets are rolled at the end of the combat move phase. But in any case scrambling has to occur before (as long as the Airfields are operative).

**Comment by Cernelius** [comment]
So, to make a summing up, assuming that here we care only about standard games, which also means games having the traditional phase order of "Combat Move" then "Conduct Combat" (but here I'll point out that also in the WWII Global map there are some mods offering the Move-Then-Buy order, that I assume are supposed to be for easy-of-play only, thus it is bad and degrading for them if they don't play as the original, rulewise):

1) There is an high level issue (major bug) that rockets can damage facilities needed to perform scramble, which should not be possible, and, in general, scramble should always happen before rocketing, while now the opposite happens.

2) There is a low level issue (minor bug) that, by rules, you should be allowed to perform rockets attacks after any bombing raids (I can see this being actually wanted), while currently you are forced to direct and resolve all rockets attack before resolving any bombing raids.

**Comment by simon33-2** [comment]
The bit about scrambling occurring after rocket attacks has been in the notes for ages.

I'm pretty sure this all can't be changed without an incompatible release and that isn't likely to be coming. BTW, why would you, as a player, resolve rockets before bombers? If a bomber is shot down, you are likely to compensate by sending a rocket there.

I don't really see the problem with the location of the scramble if the rocket firing is moved? Unless you're into the purity of having it in the correct phase.

The clarification linked above is really unfortunate. Surely it should have gone the other way and said that you had to target all your rockets before bombing anything.

**Comment by simon33-2** [comment]
I'm not sure that the scramble can be changed easily because you have to remember where the scrambles were to move the planes back.

**Comment by Cernelius** [comment]
>The clarification linked above is really unfortunate. Surely it should have gone the other way and said that you had to target all your rockets before bombing anything.

I agree, and I suggest to code this thing in a way that custom games can optionally have it work that way, as I think it is quite clear that would be more sensible.

Or, also since it is a marginal matter, it could be possible to disregard the official clarifications, but in this case it is necessary well documenting this is wanted (for sure in pos2, and maybe somewhere else too?), otherwise there is the risk that anytime in the future the matter will resurface, and someone else will "fix" it, not realizing that the behavioural difference is intended.

I also tend to believe that the rockets firing at the end of combat move, instead of at the start of combat resolution, was not an unintended bug, but something made on purpose, taking into account the standard phase order. But, you see, when stuff is not documented, everyone is left wondering what the guys that made it were thinking.

Really, a full list of all the things that are TripleA specials/exceptions/tweaks and are intended to be supported **needs** to be made, like the fact that you can do non combat moves during combat move etc..

**Comment by panther2** [comment]
> The bit about scrambling occurring after rocket attacks has been in the notes for ages.

Indeed, I am aware of that. But it sort of gets buried there. And the scramble issue is an engine related topic, worth caring about, IMHO.

**Comment by simon33-2** [comment]
The way I remember it, I programmed the fix such that for classic the rockets still fired at the end of combat movement. I can't remember a reason for that. Could just have been for compatibility, which I couldn't get to work.

I suggest the "requires major release" tag be added to this one. Refer #1674

**Comment by panther2** [comment]
> EDIT: In classic, rocket attacks occur after non combat move. An AA Gun could possibly move into range so I guess it has to be that way - it was actually non combat move rather than combat move. /EDIT

I think this is a misunderstanding, @simon33-2 

This is how rockets work in classic:

> Rocket development gives one of your antiaircraft guns rocket power once per turn. Unlike an antiaircraft gun, a rocket **can move before combat and fire during combat**. A rocket **can also move during non-combat,** if it has not just been captured, or if it has not fired. Rockets, like antiaircraft guns, cannot fire while on a transport and they are destroyed only when they are sunk on a transport. If you capture an enemy rocket, you cannot move it on the turn you capture it; also, if you do not own rocket technology yourself, then the captured rocket is considered just a regular antiaircraft gun.

**Comment by simon33-2** [comment]
Hmm, I can't see that quote in the rules I'm looking at: https://www.axisandallies.org/wp-content/uploads/2011/12/Axis-Allies-2nd-Edition.pdf

Not sure where it says they fire after non combat either. Perhaps that part was just how Triple-A did it?

**Comment by panther2** [comment]
@simon33-2 
Sorry that I forgot to mention the source of my above quote.
It is from http://www.aamc.net/bunker/aaruleclrfd.htm ,
confirmed by Krieghund.

The rulebook itself does not elaborate on the timing of the attack.

So in classic rockets fire at any point during the Combat phase.

**Comment by simon33-2** [comment]
I have trouble calling that a clarification! Triple-A doesn't implement the nomination of an AA Gun as a rocket, for example, at least to my knowledge.

Perhaps we can get rid of supporting AA Guns firing at the end of non combat though. Or was that done to simulate the ability to move an AA Gun before attacking?

**Comment by panther2** [comment]
> I have trouble calling that a clarification! 

I don't: Krieghund confirmed it is official.

> Triple-A doesn't implement the nomination of an AA Gun as a rocket, for example, at least to my knowledge.

Someone familiar with the implementation of Classic in TripleA should know. Not my expertise, here.
The rulebooks at least say so.

> Perhaps we can get rid of supporting AA Guns firing at the end of non combat though. Or was that done to simulate the ability to move an AA Gun before attacking?

Concerning A&A-games I don't see why AA-Guns (as rockets) should fire at that point. In every (world-scenario) game where rockets are included (be they AA guns or whatever) they fire during Conduct Combat Phase.

Classic: at any time during Combat
Revised (incl. LHTR): During "Conduct Opening Fire"
Anniversary: During SBR 
Global: During SBR

AA-guns "as AA-guns" fire during the Combat Move Phase of the flying-over or attacking player.

So I don't see any reason for AA or rocket fire at the end of Non Combat Move Phase - except some special maps maybe needed that implemented for specific reasons.

**Comment by simon33-2** [comment]
Thanks for that link. Hadn't seen that before. I don't know why they felt the need to retrospectively nerf rockets but not heavy bombers. I would have thought "clarifications" would be used to close loopholes and change things which were unintended and not change major game dynamics. Note that the FAQ on the official forum: www.harrisgamedesign.com/pdf/A&A_Classic_FAQ.pdf says no such thing.

Similarly the Revised FAQ upholds that multiple AA Guns can fire, only one per target though. Seems this rule is correctly implemented although I'd think you should have the option of playing the original rule or the updated rule.

Regarding the reason for firing at the end of non combat, I just explained that - AA Guns should be able to move before they fire. I wonder what was supposed to happen if you attacked with a rocket AA Gun, lost and didn't retreat?

**Comment by panther2** [comment]
> Note that the FAQ on the official forum: www.harrisgamedesign.com/pdf/A&A_Classic_FAQ.pdf says no such thing.

Yes, that one has never been updated. Both documents are valid and complement each other.

> Regarding the reason for firing at the end of non combat, I just explained that - AA Guns should be able to move before they fire.

Yes, but according to the Classic Rules they move during Combat Move Phase and fire during Combat.
Nothing moving during NCM should be allowed to fire then, at least not according to any A&A-rules.

**Comment by Cernelius** [comment]
Is classic the only game in which (only one per power) AA gun gain the ability to move during Combat Move (if they are going to fire)? Are they restricted from moving into hostile territories anyways?

**Comment by panther2** [comment]
> Is classic the only game in which (only one per power) AA gun gain the ability to move during Combat Move (if they are going to fire)?

AA guns move during NCM in all games. It's only in Classic where the "Rocket-AAgun" may move before Combat Phase.

> Are they restricted from moving into hostile territories anyways?

Good question. By the rulebook AA-guns "can be moved to other friendly land territories but never to enemy-occupied territories as an attacking land unit.". On the other hand - if the "Rocket-AA" is allowed to move before combat, that would imply the regular Combat Movement rules.

I think the clarification needs a clarification here... I am going to ask and add the answer here.

**Comment by panther2** [comment]
Here is the clarification (by Krieghund):

"It [The sentence "Unlike an antiaircraft gun, a rocket can move before combat and fire during combat."] simply means that an AA gun which will function as a rocket may move in the Combat Move phase (into a friendly territory) before firing in the Combat phase."

**Comment by simon33-2** [comment]
> To support any house rules?

For the third time, since no AA Gun can move in combat movement in Triple-A, firing at the end of NCM in classic gives a proxy for the bizarre rocket AA Gun rules.

These rules are just as bad as the G40.1 scramble rules. I'm sorry, what were they thinking? Making a retrospective change and coming up with such crap?

**Comment by Cernelius** [comment]
>These rules are just as bad as the G40.1 scramble rules. I'm sorry, what were they thinking? Making a retrospective change and coming up with such crap?

Nonetheless, these Classic rules should be the default for the program.

**Comment by panther2** [comment]
> Nonetheless, these Classic rules should be the default for the program.

@Cernelius 

Really?  The non A&A-games I came in contact with all derive from at least v2 or v3. Is everything actually based on v1? (I don't want to start a huge offtopic discussion here - I am just curious about the background of your statement.)

**Comment by Cernelius** [comment]
Almost everything is default v1. This is why you need to set the property to have v2 rules, for example.

The main exception is the naval bombard return fire, that is set default to v2 just because the property was bugged, doing the opposite of what it said, and in 1.9 @DanVanAtta changed the property name to be consistent with the behaviour (while I believe the name should have been kept and the behaviour flipped).

The other exception I know of is that you need to specifically define the territories able to place infinite units, instead of them being the existent factories, as per Classic.

**Comment by Cernelius** [comment]
In this particular case, rockets should work per v1 rules default, but having properties for setting them to the other rulesets, and such properties being comprehended in the general properties, like the one that set the game to v2 rules.

**Comment by panther2** [comment]
Interesting information. Thank you!

**Comment by ron-murhammer** [comment]
@panther2 Do you mind summarizing the issues here a bit further? My understanding from your initial report is the for Global/Pacific/Europe, that rocket attacks should occur after scrambling not before? Can you summarize how rocket attacks should work across the various A&A versions?

Here is my current understanding:
- V1 - rocket attacks happen at the end of non-combat move
- V2-V6 - rocket attacks happen at the end of combat move (I guess could happen at the start of battle phase since that would essentially be equivalent)
- Global - rocket attacks happen at the end of combat move but should happen during battle phase after scrambling

**Comment by simon33-2** [comment]
The way V1 works in Triple A is a real hack.

The way it's supposed to work is at the point of gaining the rocket tech, ONE of your AA Guns is nominated to become a rocket launcher, while still functioning as an AA Gun. This Rocket AA Gun can move on combat move to a friendly territory to fire at an industrial complex. If you lose the Rocket AA Gun, you lose the tech unless you can capture it or a different one back.

**Comment by simon33-2** [comment]
As for your comments on the long term, to be honest I would be happy if it was possible to both (a) follow the rules as clarified and (b) the rules as most people would think that they should be.

In both cases without needing edit mode.

It's not like tech is very popular.

**Comment by ron-murhammer** [comment]
I guess after reviewing the open PR to fix this. My other question is around for each A&A version what is the proper rule for target selection vs target rolling? So when should all targets be selected then all dice rolled (all target then all rolls) vs when should targets be selected and dice rolled 1 at a time (target, roll, target, roll, etc)?

**Comment by simon33-2** [comment]
@panther2 I'm wondering if we should ask Krieghund if he can reconsider his decision on this one. Would save a map property.

**Comment by panther2** [comment]
@ron-murhammer 

> @panther2 Do you mind summarizing the issues here a bit further? My understanding from your initial report is the for Global/Pacific/Europe, that rocket attacks should occur after scrambling not before? Can you summarize how rocket attacks should work across the various A&A versions?

Sorry for the delay. I hope to get this done within the next week.

**Comment by panther2** [comment]
@ron-murhammer 

**v1**
- "Choose one free rocket attack per turn! Choose one of your antiaircraft guns (it must be 3 adjacent spaces or less away from the target) to launch rockets at an enemy indusrial complex bay rolling one die. The number tossed determines how many IPC the playe owning the industrial ccomplex must surrender to the bank." (Rulebook)
- An AA-gun which will function as a rocket may move during Combat Move Phase (into a friendly territory) before firing in the Combat Phase (Clarification)
- If it has not moved during Combat Move Phase and fired during Combat it moves during Noncombat Move Phase (Clarification)

**v2** 
_to be continued_

**Comment by panther2** [comment]
**v2 OOB** 

Rocket V2 OOB

- Your antiaircraft guns are now rocket launchers. In addition to their normal combat function, they can reduce enemy industrial production. (Rulebook)
- If you have the Rockets development, your antiaircraft guns can act as rocket launchers: You can attack enemy production with your antiaircraft guns **during the opening fire step of combat**. This is the only situation in which antiaircraft guns can attack. In each territory, one antiaircraft gun may attack an industrial complex within 3 spaces, though each industrial complex may be attacked by only one
rocket launcher in a turn. The maximum loss each rocket can inflict is the target territory's income
value. The opponent must surrender that many IPCs to the bank. (Rulebook/FAQ)

- An AA gun cannot move in non­combat movement if it fired a rocket in the preceding combat phase. (FAQ)
- The target of strategic bombing or rocket attack surrenders the amount rolled by the attack or everything he has, whichever is less.

**Comment by panther2** [comment]
**v2 LHTR** 

_too be continues_

**Comment by panther2** [comment]
**v3**

- Your antiaircraft guns are now rocket launchers. In addition to its normal combat function,
during the strategic bombing raid step of your Conduct Combat phase each turn, each of your
antiaircraft guns can make a single rocket attack against an enemy industrial complex within 3 spaces
of it. This attack does 1d6 damage to that complex. In each turn, only one AA gun per territory may launch rockets, and each industrial complex can be attacked by only one rocket launcher. Rockets may not be fired over neutral territories.

I recommend to place the rocket attack (target selection + attack) at the very beginning of the Conduct Combat Phase.

**Comment by panther2** [comment]
@ron-murhammer 
> Do you mind summarizing the issues here a bit further? ...

Ready and done. :smile:

**Comment by ron-murhammer** [comment]
@panther2 Great summary. Surprised about v1 as for some reason it appears TripleA has it during NCM right now. Seems that rockets should be done at the start of conduct combat phase after scrambling across all versions. And really besides v1, there isn't as much variation as I thought.

@simon33-2 Can you review these and compare against your PR. I think we can make a few more simplifications to more closely align to these.

**Comment by Cernelius** [comment]
In v1, how about moving the AA gun with a transport, in the following cases, all done during CM, offloading to a friendly land territory, for an AA gun that it is going to rocket on the turn:
1) The AA gun was already loaded on the transport, and I offload it without moving the transport.
2) The AA gun was already loaded on the transport, and I move the transport into a hostile sea zone, and offload the AA gun (as said, the AA gun would go to a friendly land territory).
3) The AA gun was already loaded on the transport, and I move the transport into a friendly sea zone, and offload the AA gun.
4) I load the AA gun on a transport that has not moved, bridging it.
5) I load the AA gun on a transport that has not moved or at moved 1 only, and I move the transport into a hostile sea zone, and offload the AA gun.
6) I load the AA gun on a transport that has not moved or at moved 1 only, and I move the transport into a friendly sea zone or just stop its movement there, and offload the AA gun.

**Comment by simon33-2** [comment]
At first glance, I thought the case where the AA Gun was left on the transport was not properly handled but perhaps it is. If you do that, it can't fire that turn which is what I imagine it would do.

Your second case is interesting! I don't think Krieghund considered it.

Otherwise I don't think transports cause any additional problems for rockets. They're just poorly implemented in V1. I have no proposals to change this.

**Comment by panther2** [comment]
Remember that transports can only offload for an amphibious assault during combat move.

**Comment by Cernelius** [comment]
Well, that may be just a redundant clarification on how stuff works normally. Beside the rocket case, that restriction is completely superfluous, as the amphibious assault only limit is already fully dictated by the rule that land units can move during CM only to do combat or conquer territories. So, since the rockets allow to ignore that rule to some extent, the dubious if it applies to what appears to be a subset of that rule too I think cannot be clearly ruled out by the rulebook only, tho it seemingly does, literally at least.

So, practically, are you sure that the rule substantially is "_an AA-gun which will function as a rocket may move during Combat Move (into a friendly territory only) before firing in the Combat Phase (Clarification) but it still cannot offload during Combat Move_".

**Comment by panther2** [comment]
I am not going to add other potentially redundant stuff here.

**Comment by simon33-2** [comment]
@panther2 These issues should be all resolved now for all cases except v1. Which works the same as it did always.
